### ISSUE TITLE
System 573 Network Does Not Work

### ISSUE THREAD

**Issue by TRRESTGHGYTCFTGBCXSZD** [mamedev/mame#9090]
Received DNC On Both boards.
![0039](https://user-images.githubusercontent.com/40109698/148039236-aee8e596-dd3e-412d-bc46-faa1e774a284.png)
![0040](https://user-images.githubusercontent.com/40109698/148039239-40474868-de7d-43ad-8d8e-bc884e4343ac.png)

**Comment by TRRESTGHGYTCFTGBCXSZD** [comment]
Proof of this: ```
start /b mame ddrs2k -wdog 10 -nvram_directory nvram -comm_localport 15112 -comm_remoteport 15113 -comm_framesync
start /b mame ddrs2k -wdog 10 -nvram_directory nvram1 -comm_localport 15113 -comm_remoteport 15112 -comm_framesync
```

**Comment by smf-** [comment]
I don't think it's been implemented.

**Comment by Hydreigon223** [comment]
> I don't think it's been implemented.
> […](#)
> On 04/01/2022 09:41, TRRESTGHGYTCFTGBCXSZD wrote: mame ddrs2k -wdog 10 -nvram_directory nvram -comm_localport 15112 -comm_remoteport 15113 -comm_framesync

In the main branch it isn't but experimental network emulation through the digital I/O board is supported in fork by @987123879113 . With your two cfgs set up, it'd only work for two players but not more than two because of having to send a network frame/pack to all connected instances (max 4). This implementation made use of the bitbanger possibly for broadcasting sent frames to a TCP socket and then to all connected Mame instances but I'd like to know more about this. Idk if the DDR Solo games are the only games to make use of this. https://github.com/987123879113/mame/blob/bemani/src/mame/machine/k573dio.cpp

**Comment by 987123879113** [comment]
The registers are documented in my code but I didn't actually try to implement the network packets stuff in any way close to real hardware so I never tried submitting a PR. 

The RCA connectors on the back of the digital I/O board are used for the network connection, of which there are only two. According to the game manual, the 4 machine setup chains together so machine 1 connects to machine 2, machine 2 connects to machine 1 and 3, machine 3 connects to machine 2 and 4, and machine 4 connects to machine 3. It's not a loop where machine 1 or 4 connect to each other. But machines 1 and 4 can still communicate to each other so there's some packet forwarding that goes on. 

I'm not sure at what level the packets are getting forwarded (didn't seem to be game level so maybe hardware?) or when exactly they get forwarded which is where problems occurred using a pure chained bitbanger implementation. It'd end up where far nodes, like machines 1 and 4, would take too long to communicate and they'd drop connection with each other while something like machine 2 and 3 would be strong connections. 

Since I had already decided it wasn't PR worthy I made a relay server that handled packet transfer and forwarded to all connected machines at the same time and called it a day. No connection issues after that and even team play online vs other people worked perfectly.
https://gist.github.com/987123879113/f9363d18cd8947383ee5a27b6253cc81

**Comment by TRRESTGHGYTCFTGBCXSZD** [comment]
no proof of 2-4 player connected on ddr solo 2000 on youtube (shows ddr solo 2000 related but no multiplayer)

**Comment by 987123879113** [comment]
More obscure features from older games aren't all that well documented. But DDR Solo 2000 has up to 4 player network play just like DDR Solo Bass Mix.

![image](https://user-images.githubusercontent.com/63495610/148153560-a1b3cfe0-e32a-470c-8d03-267e30484d52.png)

**Comment by TRRESTGHGYTCFTGBCXSZD** [comment]
virtura formula has 8 player support
with loop directing

**Comment by rb6502** [comment]
Virtua Formula is a different game on different hardware in a different genre from a different company.  I'm not sure what it has to do with DDR.

**Comment by Hydreigon223** [comment]
They connect over RS485 (in terms of emulation, this would mean a network packet would have to be sent to three mame instances at once and possibly itself in the process).

**Comment by 987123879113** [comment]
>in terms of emulation, this would mean a network packet would have to be sent to three mame instances at once and possibly itself in the process

As per the wiring diagram, it's only ever possible to talk to up to 2 machines at once so it'd be impossible to send the same packet to 3 MAME instances at once. It'd have to be chained around between machines to reach machines not directly connected. The implementation only became a problem when the indirectly connected machines are involved. 2 machines connected together don't require filtering or rebroadcasting so is a much simpler case.

I took a look at the mentioned implementations in nwk-tr and k056230 and while they seem similar enough, it doesn't look like those have the code required to connect multiple instances of MAME together so I have no idea if those actually work properly when networked together.

**Comment by Hydreigon223** [comment]
> > in terms of emulation, this would mean a network packet would have to be sent to three mame instances at once and possibly itself in the process
> 
> As per the wiring diagram, it's only ever possible to talk to up to 2 machines at once so it'd be impossible to send the same packet to 3 MAME instances at once.

That's what I thought at first...until I figured the gist of RS485 and verifying continuity with both RCA network ports on a GTI Club pcb. The center conductors on both ports share continuity as with the outer conductors respectively. It should be the same with the digital I/O board (I suggest checking both ports with a multimeter).

**Comment by 987123879113** [comment]
I read up on it some more and I think I understand what you are saying now. 

On Windows anyway, it's not possible to reuse sockets between the different MAME instances and it will throw the error "failed: address in use" if you try, so you'd need 3 connections per machine (6 unique sockets total) to get all machines connected directly to each other. It's simple to modify my current code to use 3 sockets and remove the few lines that are rebroadcasting packets so that part isn't an issue, but it's not user friendly having to work out how to connect the machines properly.

```
mame.exe -w ddrsbm -bitbanger1 socket.127.0.0.1:5731 -bitbanger2 socket.127.0.0.1:5732 -bitbanger3 socket.127.0.0.1:5733
mame.exe -w ddrsbm -bitbanger1 socket.127.0.0.1:5731 -bitbanger2 socket.127.0.0.1:5734 -bitbanger3 socket.127.0.0.1:5735
mame.exe -w ddrsbm -bitbanger1 socket.127.0.0.1:5733 -bitbanger2 socket.127.0.0.1:5734 -bitbanger3 socket.127.0.0.1:5736
mame.exe -w ddrsbm -bitbanger1 socket.127.0.0.1:5732 -bitbanger2 socket.127.0.0.1:5735 -bitbanger3 socket.127.0.0.1:5736
```

Something also needs to be done for flow control, which I hacked around in my implementation by detecting full packets instead, or else the clients will all be sending data at the exact same time resulting in jumbled packets everywhere. I'm not really sure how that would yet.

The relay server solution is feeling a lot simpler and cleaner in comparison now.

**Comment by TRRESTGHGYTCFTGBCXSZD** [comment]
can you show up ddr solo 2000 (asia) multiplayer in video @987123879113

**Comment by 987123879113** [comment]
No. It is literally the same exact thing as playing it single player, except you see scores for the other players on the result screen. There's no reason to record a gameplay video because it's exactly the same as normal.

**Comment by TRRESTGHGYTCFTGBCXSZD** [comment]
Found it. There's Split Second Playing Multiplayer.
https://www.youtube.com/watch?v=KwKW-BmyKrU
